lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
-          if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
+          unless serializer.class._type.nil?
+            serializer.class._type
           else
-            serializer.object.class.model_name.plural
+            if ActiveModel::Serializer.config.type_source == :model
+              type = serializer.object.class.name.underscore
+            else
+              type = serializer.class.name.underscore.split('_serializer').first
```

Comment:
This seems a bit overcomplicated. How about `serializer.class.name.chomp('Serializer').underscore`?